lib/metrics.js

```diff
@@ -0,0 +1,848 @@
+ * - Timer: A duration in milliseconds.
+ * - UniqueSet: A unique count of number of unique values seen.
+ * - PeriodicGauge: A gauge which periodically updates its value by calling a function.
+ *
+ * # Exporting Metrics
+ *
+ * Several text format exporters are provided as streams:
+ * - Statsd
+ * - Dogstatsd
+ * - Graphite
+ * - Prometheus
+ *
+ * # TODO(qard):
+ * - Histograms
+ * - Distributions/Summaries
+ */
+
+'use strict';
+
+const {
+  ArrayPrototypeJoin,
+  ArrayPrototypeMap,
+  MathFloor,
+  ObjectAssign,
+  ObjectEntries,
+  ObjectFreeze,
+  ObjectKeys,
+  SafeMap,
+  SafeWeakSet,
+  SymbolDispose,
+} = primordials;
+
+const {
+  codes: {
+    ERR_INVALID_ARG_TYPE,
+    ERR_INVALID_ARG_VALUE,
+  },
+} = require('internal/errors');
+const { setInterval, clearInterval } = require('internal/timers');
+
+const {
+  channel,
+  hasChannel,
+  subscribe,
+  unsubscribe,
+} = require('diagnostics_channel');
+const { performance } = require('perf_hooks');
+const { Transform } = require('stream');
+
+const newMetricChannel = channel('metrics:new');
+
+/**
+ * Mix two metadata objects together.
+ * @param {object} a The first metadata object.
+ * @param {object} b The second metadata object.
+ * @returns {object} The mixed metadata.
+ * @private
+ */
+function mixMeta(a, b) {
+  if (a === undefined) return b;
+  if (b === undefined) return a;
+  return ObjectAssign({}, a, b);
+}
+
+/**
+ * Represents a single reported metric.
+ */
+class MetricReport {
+  /**
+   * The type of metric.
+   * @property {string} type
+   */
+
+  /**
+   * The name of the metric.
+   * @property {string} name
+   */
+
+  /**
+   * The value of the metric.
+   * @property {number} value
+   */
+
+  /**
+   * Additional metadata to include with the report.
+   * @property {object} meta
+   */
+
+  /**
+   * Constructs a new metric report.
+   * @param {string} type The type of metric.
+   * @param {string} name The name of the metric.
+   * @param {number} value The value of the metric.
+   * @param {object} [meta] Additional metadata to include with the report.
+   */
+  constructor(type, name, value, meta) {
+    this.type = type;
+    this.name = name;
+    this.value = value;
+    this.meta = meta;
+    this.time = performance.now();
+    ObjectFreeze(this);
+  }
+
+  /**
+   * Convert the metric report to a statsd-compatible string.
+   * @returns {string} The statsd-formatted metric report.
+   */
+  toStatsd() {
+    const { type, name, value } = this;
+    return `${name}:${value}|${this.#statsdType(type)}`;
+  }
+
+  /*
+   * Convert the metric type to a statsd type.
+   *
+   * @param {string} type The metric type.
+   * @returns {string} The statsd type.
+   * @private
+   */
+  #statsdType(type) {
+    return {
+      counter: 'c',
+      gauge: 'g',
+      meter: 'm',
+      periodicGauge: 'g',
+      timer: 'ms',
+      uniqueSet: 's',
+    }[type];
+  }
+
+  /**
+   * Convert the metric report to a Dogstatsd-compatible string.
```

Comment:
Would be good to include links to the relevant specs for these in the comments for future generations :-)

Reply:
Assuming the formatting bits remain the the final design, certainly. 🙂